### DISCUSSION TITLE
Is MSP support a thing?

### DISCUSSION THREAD

**Discussion by celtichawk** [Blightmud/Blightmud#277] [Q&A]
First of, osry if this has been asked  lot before, I've not ben able to search, but I came across Blightmud since I'm a TT++ user mostly on Ubuntu and I like what I'm seeing with Blightmud, however I'm unsure if it has/is planned to have any MSP support or not for Linux?

If not currently, is it planned to? Again sory if this is the wrong place, but...

**Comment by LiquidityC** [comment]
Hello, yes. MSP is a thing but not in core yet. There are two issues that are "hodling it back".
1. The audio library we use `rodio` has some quirks (with an active PR) that I'm waiting for
2. I have a lua MSP implementation which isn't complete but I have been using it a lot. This is what I'm planning on releasing as a plugin to start.

If you are building Blightmud from the `dev` branch then audio works just fine. If you want to try your hand on the MSP plugin I can give you what I have so far. I have yet to test a mud that extensively implements MSP hence the limitations on my version. I also think we will need one or two more features within Blightmuds core audio to fully support it. My original plan was to ship a complete set with version 3.1. Now I'm not certain due to the state of `rodio`. I hope this answer helps you a bit. Feel free to follow up, either here or on our Discord.

```lua
-- Blightmud MSP support
local mod = {}

local audio_dir = nil
local filetype = "mp3"

-- Call this to set the audio dir
function mod.set_dir(dir)
	core.enable_protocol(90)
	audio_dir = dir
end

-- Call this to set the audio filetype
function mod.set_filetype(suffix)
	filetype = suffix
end

local function file_exists(file)
	local f = io.open(file)
	if f then
		f:close()
		return true
	else
		return false
	end
end

local function split(inputstr, sep)
	if sep == nil then
		sep = "%s"
	end
	local t={}
	for str in string.gmatch(inputstr, "([^"..sep.."]+)") do
		table.insert(t, str)
	end
	return t
end

local function get_params(input)
	local inputs = split(input, " ")
	local params = {}
	local sound = table.remove(inputs, 1)
	params["sound"] = sound
	for _,v in ipairs(inputs) do
		local key, value = table.unpack(split(v, "="))
		params[key] = value
	end
	return params
end

local triggers = trigger.add_group()
triggers:add("^!!SOUND\\(([\\w\\d= ]+)\\)", {gag=true, raw=true},
	function (m)
		if audio_dir then
			local params = get_params(m[2])
			local sound_file = audio_dir .. "/" .. params["sound"] .. "." .. filetype
			if file_exists(sound_file) then
				audio.play_sfx(sound_file)
			else
				info("MSP", "No sound for: " .. params["sound"])
			end
		end
	end
)

triggers:add("^!!MUSIC\\(([\\w\\d= ]+)\\)", {raw=true, gag=true},
	function (m)
		if audio_dir then
			local params = get_params(m[2])
			local sound_file = audio_dir .. "/" .. params["sound"] .. "." .. filetype
			if file_exists(sound_file) then
				audio.play_music(sound_file)
			else
				info("MSP", "No sound for: " .. params["sound"])
			end
		end
	end
)

return mod
```

**Reply by LiquidityC** [reply]
I moved what I had to an actual plugin. It's not ready but it's something to work off of: https://github.com/Blightmud/msp

**Comment by celtichawk** [comment]
Ah that helps a bit, I do have a MUD I can test it on, I can see how this goes with the lua you provided, I'm unsure however if the audio will work cros cross platform (I'm building it on Ubuntu here)

**Comment by LiquidityC** [comment]
Linux is our primary target platform. You shouldn't have any issues.
